docs/design-documents/features/storage/SecureStore/SecureStore_design.md

```diff
@@ -0,0 +1,448 @@
+# SecureStore in Mbed OS
+
+- [SecureStore in Mbed OS](#securestore-in-mbed-os)
+    + [Revision history](#revision-history)
+- [Introduction](#introduction)
+    + [Overview and background](#overview-and-background)
+    + [Requirements and assumptions](#requirements-and-assumptions)
+- [System architecture and high-level design](#system-architecture-and-high-level-design)
+  * [Design basics](#design-basics)
+    + [Data layout](#data-layout)
+    + [Basic implementation concepts](#basic-implementation-concepts)
+- [Detailed design](#detailed-design)
+    + [Class header](#class-header)
+    + [Important data structures](#important-data-structures)
+    + [Initialization and reset](#initialization-and-reset)
+    + [Core APIs](#core-apis)
+    + [Incremental set APIs](#incremental-set-apis)
+    + [Key iterator APIs](#key-iterator-apis)
+- [Usage scenarios and examples](#usage-scenarios-and-examples)
+    + [Standard usage of the class](#standard-usage-of-the-class)
+- [Other information](#other-information)
+    + [Open issues](#open-issues)
+
+
+### Revision history
+
+| Revision 	| Date           	| Authors                                                	| Mbed OS version 	| Comments         	|
+|----------	|----------------	|--------------------------------------------------------	|-----------------	|------------------	|
+| 1.0      	| 02 October 2018	| David Saada ([@davidsaada](https://github.com/davidsaada/)) 	| 5.11+           	| Initial revision 	|
+
+# Introduction
+
+### Overview and background
+
+SecureStore is a [KVStore](../KVStore/KVStore_design.md) based storage solution, providing security features on the stored data, such as encryption, authentication, rollback protection and write once, over an underlying KVStore class. It references an additional KVStore class for storing the rollback protection keys. 
+
+### Requirements and assumptions
+
+SecureStore assumes that the underlying KVStore instances are instantiated and initialized. 
```

Comment:
I would expect this to match the behaviour of FileSystem -> BlockDevice. Actually this is related to above comment.

Reply:
Please see my comment above.

Reply:
I'm still expecting SecureStore to match the behaviour of SlicingBlockDevice -> BlockDevice.

Consider:
``` cpp
TDBStore kv;
SecureStore secure_kv(&kv);
```

Do we need multiple init calls in this case? Does this mean the target needs to call all inits in the storage stack before providing KVStore::get_default_instance?

Reply:
Didn't know we have `get_default_instance` API for KVStore - maybe we should add it. Regardless, instantiating the KVStore class is something that should be implemented by the application init code (what we call the "integration code"). This code should instantiate the block device(s) and file system(s), so having an additional init call doesn't seem excessive. However, it's not `init` that bothers me like `deinit`. While calling file system/block device `init` by KVStore may be harmless calling `deinit` automatically may be more dangerous if the file system/block device is used for other purposes (which also relates to our discussion above). This was partially solved for block devices (with the init ref count), but will also need to get resolved for file systems if we follow this approach.

Reply:
Irrespective of if we chain inits or not, we should have consistent API patterns in BlockDevice, KVStore, and (eventually) FileSystem.

Currently standardizing on chaining inits seems like the easiest option.

I'm not sure how we could move away from chained inits in BlockDevice without breaking user's code.

---

> Didn't know we have get_default_instance API for KVStore - maybe we should add it. 

Well we have FileSystem::get_default_instance and BlockDevice::get_default_instance so it sounds like a good idea : )

> This code should instantiate the block device(s) and file system(s), so having an additional init call doesn't seem excessive. 

get_default_instance is similar to class constructors, it doesn't return errors. So we should be avoiding init calls and leaving that up to the user.

The current behavior of BlockDevice::get_default_instance does this:
``` cpp
BlockDevice *bd = BlockDevice::get_default_instance();
bd->init(); // must call init
```

The behavior for FileSystem::get_default_instance is weird, just like the FileSystem constructor. I think we should improve this in the future.

---

Is there any issues with using the ref count in KVStore and FileSystem? This provides the most resilience to incorrect init calls from users.

Reply:
I tend to agree with everything you write here. There's no real issue with implementing init ref count   on file systems, other than time constrains. This is probably no biggie. Real problem is mainly with `deinit`, which isn't really used in real life, but more on test, so we may postpone it to 5.12.